xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,291 @@
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "CDS load balancer is shutting down");
+
+    switchingLoadBalancer.shutdown();
+    if (xdsClientRef != null) {
+      xdsClientRef.returnObject(xdsClient);
+    }
+  }
+
+  /**
+   * A LoadBalancerProvider that provides a load balancer with a fixed CdsConfig.
+   */
+  private final class FixedCdsConfigBalancerProvider extends LoadBalancerProvider {
+
+    final CdsConfig cdsConfig;
+    final CdsConfig oldCdsConfig;
+    final ClusterWatcher oldClusterWatcher;
+
+    FixedCdsConfigBalancerProvider(CdsConfig cdsConfig) {
+      this.cdsConfig = cdsConfig;
+      oldCdsConfig = CdsLoadBalancer.this.cdsConfig;
+      oldClusterWatcher = CdsLoadBalancer.this.clusterWatcher;
+    }
+
+    @Override
+    public boolean isAvailable() {
+      return true;
+    }
+
+    @Override
+    public int getPriority() {
+      return 5;
+    }
+
+    // A synthetic policy name identified by CDS config.
+    @Override
+    public String getPolicyName() {
+      return "cds_policy__cluster_name_" + cdsConfig.name;
+    }
+
+    @Override
+    public LoadBalancer newLoadBalancer(final Helper helper) {
+      return new LoadBalancer() {
+        // Becomes non-null once handleResolvedAddresses() successfully.
+        // Assigned at most once.
+        @Nullable
+        ClusterWatcherImpl clusterWatcher;
+
+        @Override
+        public void handleNameResolutionError(Status error) {
+          if (clusterWatcher == null || clusterWatcher.edsBalancer == null) {
+            // Go into TRANSIENT_FAILURE if we have not yet received any cluster resource.
+            // Otherwise, we keep running with the data we had previously.
+            helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+          }
+        }
+
+        @Override
+        public boolean canHandleEmptyAddressListFromNameResolution() {
+          return true;
+        }
+
+        @Override
+        public void shutdown() {
+          if (clusterWatcher != null) {
+            if (clusterWatcher.edsBalancer != null) {
+              clusterWatcher.edsBalancer.shutdown();
+            }
+            xdsClient.cancelClusterDataWatch(cdsConfig.name, clusterWatcher);
+          }
+        }
+
+        @Override
+        public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+          if (clusterWatcher == null) {
+            clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+            xdsClient.watchClusterData(cdsConfig.name, clusterWatcher);
+            if (oldCdsConfig != null) {
+              xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+            }
+            CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+          }
+          // Else handleResolvedAddresses() already called, so no-op here because the config is
```

Comment:
This comment does not say anything. Code already explains that only first call of `handleResolvedAddresses(...)` sets `clusterWatcher`. This sentence is saying that "config is fixed in FixedCdsConfigBalancer", which literally means nothing...

Whatever...

Reply:
Whatever...  removed.